twingate/client_resource.go

```diff
@@ -301,20 +293,67 @@ func (client *Client) readResourceGroupsAfter(ctx context.Context, resourceID gr
 
 type readResourcesQuery struct { //nolint
 	Resources struct {
-		Edges []*Edges
+		PageInfo PageInfo
+		Edges    []*Edges
 	}
 }
 
 func (client *Client) readResources(ctx context.Context) ([]*Edges, error) { //nolint
 	response := readResourcesQuery{}
-	variables := map[string]interface{}{}
 
-	err := client.GraphqlClient.NamedQuery(ctx, "readResources", &response, variables)
+	err := client.GraphqlClient.NamedQuery(ctx, "readResources", &response, nil)
 	if err != nil {
 		return nil, NewAPIErrorWithID(err, "read", resourceResourceName, "All")
 	}
 
-	return response.Resources.Edges, nil
+	edges, err := client.readAllResources(ctx, &response)
+	if err != nil {
+		return nil, err
+	}
+
+	return edges, nil
+}
+
+func (client *Client) readAllResources(ctx context.Context, resource *readResourcesQuery) ([]*Edges, error) { //nolint
```

Comment:
Is there a reason to separate this to its own func and not have it as part of `readResources` ?
Having this receive the previous response (`resource *readResourcesQuery`) makes it not really an independent function...

Reply:
it's more like private function, separated it for readability purpose

Reply:
maybe I can define this func on query object and pass client as argument

Reply:
I guess moving func to another object doesn't make much sense
so you proposing to merge this functions like this, right?

```

type ResourcesEdges struct {
	PageInfo PageInfo
	Edges    []*Edges
}

type readResourcesQuery struct { //nolint
	Resources ResourcesEdges
}

func (client *Client) readResources(ctx context.Context) ([]*Edges, error) { //nolint
	response := readResourcesQuery{}

	err := client.GraphqlClient.NamedQuery(ctx, "readResources", &response, nil)
	if err != nil {
		return nil, NewAPIErrorWithID(err, "read", resourceResourceName, "All")
	}

	resource := &response.Resources
	edges := resource.Edges
	page := resource.PageInfo

	for page.HasNextPage {
		resp, err := client.readResourcesAfter(ctx, page.EndCursor)
		if err != nil {
			return nil, err
		}

		edges = append(edges, resp.Edges...)
		page = resp.PageInfo
	}

	return edges, nil
}

type readResourcesAfterQuery struct { //nolint
	Resources ResourcesEdges `graphql:"resources(after: $resourcesEndCursor)"`
}

func (client *Client) readResourcesAfter(ctx context.Context, cursor graphql.String) (*ResourcesEdges, error) { //nolint
	response := readResourcesAfterQuery{}
	variables := map[string]interface{}{
		"resourcesEndCursor": cursor,
	}

	err := client.GraphqlClient.NamedQuery(ctx, "readResource", &response, variables)
	if err != nil {
		return nil, NewAPIError(err, "read", resourceResourceName)
	}

	if len(response.Resources.Edges) == 0 {
		return nil, NewAPIError(ErrGraphqlResultIsEmpty, "read", resourceResourceName)
	}

	return &response.Resources, nil
}
```

Reply:
but it looks that in `readResources` func we operating on different abstraction level
we can improve it like this I think

@ekampf what do you think?

```
func (client *Client) readResources(ctx context.Context) ([]*Edges, error) { //nolint
	//response := readResourcesQuery{}
	//
	//err := client.GraphqlClient.NamedQuery(ctx, "readResources", &response, nil)
	//if err != nil {
	//	return nil, NewAPIErrorWithID(err, "read", resourceResourceName, "All")
	//}
	//resource := &response.Resources

	resource, err := client.getResourcesFirstPage(ctx)
	if err != nil {
		return nil, err
	}

	edges := resource.Edges
	page := resource.PageInfo

	for page.HasNextPage {
		resp, err := client.readResourcesAfter(ctx, page.EndCursor)
		if err != nil {
			return nil, err
		}

		edges = append(edges, resp.Edges...)
		page = resp.PageInfo
	}

	return edges, nil
}

func (client *Client) getResourcesFirstPage(ctx context.Context) (*ResourcesEdges, error) {
	response := readResourcesQuery{}

	err := client.GraphqlClient.NamedQuery(ctx, "readResources", &response, nil)
	if err != nil {
		return nil, NewAPIErrorWithID(err, "read", resourceResourceName, "All")
	}

	return &response.Resources, nil
}
```

Reply:
@ekampf what do you think?

Reply:
@ekampf I've updated other resources with multiple pages as well